onnxruntime/core/platform/posix/env.cc

```diff
@@ -263,15 +275,60 @@ class PosixEnv : public Env {
   }
 
   int GetNumCpuCores() const override {
-    // TODO if you need the number of physical cores you'll need to parse
-    // /proc/cpuinfo and grep for "cpu cores".
-    // However, that information is not always available(output of 'grep -i core /proc/cpuinfo' is empty)
-    return std::thread::hardware_concurrency();
-  }
+#ifdef CPUINFO_SUPPORTED
+    if(cpuinfo_available_) {
+      return gsl::narrow<int>(cpuinfo_get_cores_count());
+    }
+#endif
+    return std::max(1, static_cast<int>(std::thread::hardware_concurrency() / 2));
+  }
+
+  std::vector<ThreadOptions::ThreadAffinity> GetThreadAffinityMasks() const override {
+    auto generate_vector_of_n = [](unsigned int num_logical_proc) {
+      std::vector<ThreadOptions::ThreadAffinity> ret;
+      ret.reserve(num_logical_proc / 2);
+      num_logical_proc = std::max(1U, num_logical_proc);
+      if (num_logical_proc == 1U) {
+        ThreadOptions::ThreadAffinity th_aff{{0}};
+        ret.push_back(std::move(th_aff));
+        return ret;
+      }
+      for (unsigned int c = 0; c + 1 < num_logical_proc; c += 2) {
+        ThreadOptions::ThreadAffinity th_aff{{static_cast<int>(c), static_cast<int>(c + 1)}};
+        ret.push_back(std::move(th_aff));
+      }
+      return ret;
+    };
 
-  std::vector<size_t> GetThreadAffinityMasks() const override {
-    std::vector<size_t> ret(std::thread::hardware_concurrency() / 2);
-    std::iota(ret.begin(), ret.end(), 0);
+    std::vector<ThreadOptions::ThreadAffinity> ret;
+#ifdef CPUINFO_SUPPORTED
+    if (cpuinfo_available_) {
+#if !defined(__APPLE__) && !defined(__ANDROID__) && !defined(__wasm__) && !defined(_AIX)
+      // We currently do not implement affinity on more than 64 cores.
+      auto num_phys_cores = cpuinfo_get_cores_count();
+      ret.reserve(num_phys_cores);
+      for (uint32_t i = 0; i < num_phys_cores; ++i) {
+        const auto* core = cpuinfo_get_core(i);
+        ThreadOptions::ThreadAffinity th_aff;
+        // Processor count will never exceed 64 in a given group.
+        // TBD: Processor groups are currently not taken into account.
+        th_aff.logical_proc_ids.reserve(core->processor_count);
+        auto log_proc_idx = core->processor_start;
+        for (uint32_t count = 0; count < core->processor_count; count++, ++log_proc_idx) {
+          const auto* log_proc = cpuinfo_get_processor(log_proc_idx);
```

Comment:
>cpuinfo_get_processor [](http://example.com/codeflow?start=33&length=21)

Will cpu_info_... calls throw? Should we place some exception handling here?

Reply:
C functions do not throw.

Reply:
What would you place in the handler if it did?